Content.Server/Database/ServerDbManager.cs

```diff
@@ -86,131 +48,360 @@ public void Init()
+
+            var prefs = await ServerDbContext.Preferences
+                .Where(p => p.UserId == userId)
+                .SingleOrDefaultAsync();
+
+            // To explain: The code later will return null if it encounters preferences and no profiles.
+            // This is because the preferences are corrupt, and the DB code doesn't have access to the function to create a new profile.
+            // In this case, a Preference object already exists but is safe to reuse.
+            if (prefs is null)
+            {
+                prefs = new Preference
+                {
+                    UserId = userId.UserId
+                };
+                ServerDbContext.Preferences.Add(prefs);
+                Logger.WarningS("c.s.db.serverdbmanager", "Reinitializing preferences[{0}] (user {1}) - this only happens in erroneous situations.", prefs.Id, userId);
+            }
+
+            ServerDbContext.PreferenceProfiles
+                .Add(new PreferenceProfile {
+                    Preference = prefs,
+                    Profile = profile
+                });
+
+            prefs.Profiles = new List<Profile> { profile };
+
+            await ServerDbContext.SaveChangesAsync();
+
+            return new PlayerPreferences(new[] {new KeyValuePair<int, ICharacterProfile>(0, defaultProfile)}, 0);
         }
 
-        public Task SaveSelectedCharacterIndexAsync(NetUserId userId, int index)
+        public async Task SaveSelectedCharacterIndexAsync(NetUserId userId, int index)
         {
-            return _db.SaveSelectedCharacterIndexAsync(userId, index);
+            var preference = await ServerDbContext.Preferences
+                .Where(p => p.UserId == userId)
+                .SingleAsync();
+
+            var selected = await ServerDbContext.PreferenceProfiles
+                .Where(p => p.PreferenceId == preference.Id)
+                .SingleOrDefaultAsync();
+
+            var profile = await ServerDbContext.Profiles
+                .Where(p => p.Slot == index && p.PreferenceId == preference.Id)
+                .SingleOrDefaultAsync();
+
+            if (selected != null)
+                ServerDbContext.Remove(selected);
+
+            if (profile != null)
+                ServerDbContext.Add(new PreferenceProfile {
+                    PreferenceId = preference.Id,
+                    ProfileId = profile.Id
+                });
+
+            await ServerDbContext.SaveChangesAsync();
         }
 
-        public Task SaveCharacterSlotAsync(NetUserId userId, ICharacterProfile? profile, int slot)
+        public async Task SaveCharacterSlotAsync(NetUserId userId, ICharacterProfile? profile, int slot)
         {
-            return _db.SaveCharacterSlotAsync(userId, profile, slot);
+            if (profile is null)
+            {
+                var profilesToDelete = ServerDbContext.Preferences
+                    .Single(p => p.UserId == userId.UserId)
```

Comment:
This should be fixed to be properly asynchronous (yes I realize the code before did this as well).